File: 2025/3/main.py
#!/usr/bin/env python3
import sys
sys.path.append('../..')
from py_utils.grid import *
from py_utils.fast_parse import *
from py_utils.search import *

from collections import defaultdict

import fileinput
import itertools
import math
import os
import queue
import re

def dprint(*args, **kwargs):
    if True:
        return print(*args, **kwargs)

p1 = None
p2 = None

def solve(lines, battery_count):
    r = 0
    for line in lines:
        num_str = ""
        start = 0
        for num in range(0, battery_count):
            largest_digit = line[start]
            start += 1
            end = min(len(line), num + len(line) - battery_count  + 1)
            for i in range(start, end):
                c = line[i]
                if c > largest_digit:
                    largest_digit = c
                    start = i+1
            num_str += largest_digit
        print(num_str)
        r += int(num_str)
    return r

def solve1(lines):
    return solve(lines, 2)

def solve2(lines):
    return solve(lines, 12)

if __name__ == "__main__":
    f = fileinput.input()
    lines = []
    for line in f:
        if line[-1] == '\n':
            line = line[:-1]
        lines.append(line)
    p1 = solve1(lines)
    p2 = solve2(lines)
    if p1 is not None:
        print("Solution 1:", p1)
    if p2 is not None:
        print("Solution 2:", p2)

File: 2025/3/main.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define LINE_SIZE 1024
#define MAX_LINES 5000
typedef char lines_t[MAX_LINES][LINE_SIZE];

long long solve(int line_count, lines_t lines, int battery_count) {
  long long result = 0;
  for (int line_index = 0; line_index < line_count; line_index++) {
      char num_str[LINE_SIZE] = "";
      char *line = lines[line_index];
      if (strlen(line) < battery_count) {
          continue;
      }
      int start = 0;
      for (int battery_n = 0; battery_n < battery_count; battery_n++) {
          char largest_digit = line[start];
          assert(start < (int)strlen(line));
          start += 1;
          int end = (int)strlen(line) - battery_count + battery_n;
          if (end > (int)strlen(line)) {
              end = (int)strlen(line);
          }
          for (int i = start; i < end; i++) {
              char c = line[i];
              if (c > largest_digit) {
                  largest_digit = c;
                  start = i + 1;
              }
          }
          num_str[battery_n] = largest_digit;
      }
      num_str[battery_count] = '\0';
      result += atoll(num_str);
  }
  return result;
}

long long solve1(int line_count, lines_t lines) {
  return solve(line_count, lines, 2);
}

long long solve2(int line_count, lines_t lines) {
  return solve(line_count, lines, 12);
}

int main () {
  lines_t lines;
  int i = 0;
  while (fgets(lines[i++], LINE_SIZE, stdin)) {
    // printf("%s", lines[i-1]);
  }
  const long long solution1 = solve1(i, lines);
  if (solution1 != 0) {
    printf("Solution 1: %lld\n", solution1);
  }
  const long long solution2 = solve2(i, lines);
  if (solution2 != 0) {
    printf("Solution 2: %lld\n", solution2);
  }
}
